### ISSUE TITLE
TypeScript SDK installer doesn't install targets for BuildTools 

### ISSUE THREAD

**Issue by stffabi** [Microsoft/TypeScript#16758]
**TypeScript Version:**  SDK installer 2.3.3

**Prerequisites:** Machine with only the [BuildTools for Visual Studio 2017](https://www.visualstudio.com/downloads/#build-tools-for-visual-studio-2017) installed.

**Expected behavior:**

Installing the TypeScript SDK on a machine with only BuildTools 2017 intalled, the SDK installer should also install the needed msbuild targets and props under the folder `C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\TypeScript`. The SDK itself is installed as expected in the folder `C:\Program Files (x86)\Microsoft SDKs\TypeScript`.

**Actual behavior:**

The msbuild files under  `C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\TypeScript` are missing. It seems like currently these are only installed if a full VisualStudio installation could be found.

**Comment by stffabi** [comment]
The problem still exists, even with TypeScript 2.4.1 SDK.

**Comment by mhegazy** [comment]
have you tried installing the [VS 2015 installer](https://www.microsoft.com/en-us/download/details.aspx?id=48593), it will install the tools even if VS is not installed on the machine.
Also consider using [the TS Nuget package](https://github.com/Microsoft/TypeScript/wiki/Configuring-MSBuild-projects-to-use-NuGet) instead.

**Comment by stffabi** [comment]
Well I've tried your mentioned VS2015 installer, but it doesn't change anything. As said before the SDK is installed, but not the targets under `C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\TypeScript` where msbuild is going to search for building the typescript project.

It behaves exactly the same way as the [VS2017 installer](https://www.microsoft.com/en-us/download/details.aspx?id=55258).

Using the nuget package is going to help, but I find it quiet not intuitive, especially one has to remove some imports manually. Well I know that's by design how nuget works 😄 . As that is only a one time thing, that shouldn't be a big problem. 
But what is VisualStudio going to do when TS is installed through nuget? Let's say we have TS 2.4.1 installed as a nuget package, but on the machine (through the TS installer) there's only TS 2.3 installed. So building through msbuild or from VisualStudio TS 2.4 will be used, but what is VisualStudio going to use for intellisense and background compilation? Will it also use TS from the nuget package?

**Comment by minestarks** [comment]
@stffabi Are you running 15.3? TypeScript 2.3 (and higher SDKs) requires Visual Studio 2017 version 15.3. 

You're correct to point out that the SDK installer doesn't update the targets files under the VS directory, but with 15.3, the targets files under the VS directory should be locating and importing the latest installed TypeScript files on your machine (in this case 2.3).

**Comment by minestarks** [comment]
> But what is VisualStudio going to do when TS is installed through nuget? Let's say we have TS 2.4 installed as a nuget package, but on the machine (through the TS installer) there's only TS 2.3 installed. So building through msbuild or from VisualStudio TS 2.4 will be used, but what is VisualStudio going to use for intellisense and background compilation? Will it also use TS from the nuget package?

Visual Studio will not use the nuget package for Intellisense -- it will use what's on the machine (2.3 in your example). So yes for full Intellisense support you'll want the SDK installed.

**Comment by stffabi** [comment]
@minestarks I'm trying to get TS run with Visual Studio 2017 BuildTools (15.3) only installed. I would like to only install the build-tools on our build servers and not the complete VisualStudio 2017 (15.3). The problem is that the SDK installer doesn't install the targets (which are needed to locate and import the correct TS version files) under `Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\TypeScript` on build-tools only machines. Nor does the build tools installer install the "locating" targets under the above directory.

But I think we are going to use the nuget packages, which will fix the problem for our build-servers. For the developer experience we are going to additionally install the SDKs on the developer machines.

**Comment by minestarks** [comment]
Something still isn't adding up if this scenario isn't working for you without Nuget. The SDK installer doesn't place anything under the Visual Studio folder, but the build tools scenario is still expected to work. 

> The problem is that the SDK installer doesn't install the targets (which are needed to locate and import the correct TS version files) under Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\TypeScript on build-tools only machines. 

Expected.

> Nor does the build tools installer install the "locating" targets under the above directory.

Not expected. To confirm:

Does your `Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\TypeScript\Microsoft.TypeScript.targets` file have this line?

`<Import Project="$(TypeScriptSdkDir)\versions\*" />`

This is the part that goes looking for the installed TypeScript SDKs under `C:\Program Files (x86)\Microsoft SDKs\TypeScript` . If it's not there then this may be a setup problem on VS's part, I'd need to check on that.

If it's there, then we should be locating the installed TypeScript SDKs and selecting the highest one by default, unless you specified a `TypeScriptToolsVersion` in your project that's different.

**Comment by stffabi** [comment]
> > Nor does the build tools installer install the "locating" targets under the above directory.
> Not expected. To confirm:
>
>Does your Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\TypeScript\Microsoft.TypeScript.targets file have this line?

No, the folder `Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\TypeScript\` is completely missing. I've installed BuildTools `15.3.26730.8` with the components `Microsoft.VisualStudio.Workload.MSBuildTools, Microsoft.VisualStudio.Workload.WebBuildTools, Microsoft.VisualStudio.Component.Static.Analysis.Tools`. Do I need an additional component for these targets to be installed?

**Comment by minestarks** [comment]
Thanks @stffabi for the very detailed info. As you've probably figured out by now, the problem is that the Build Tools VS setup graph is missing the necessary TypeScript components -- sorry about that. I'll look into getting them added. I think your NuGet workaround is the most sensible solution for now, until we get VS build tools setup fixed. 

Thanks for helping me track this down!

**Comment by stffabi** [comment]
Does anyone know if this has been fixed with the BuildTools for [Visual Studio 2017 15.4](https://www.visualstudio.com/en-us/news/releasenotes/vs2017-relnotes)?

**Comment by billti** [comment]
It has not been added to the BuildTools yet.

**Comment by pascalberger** [comment]
@minestarks @billti Any plans when this will be fixed?

**Comment by billti** [comment]
It will not be a part of the upcoming VS 2017 Update 5 releases, which means at least a few months away. I'm working on it.... we just don't have a solution ready in time for the next release.

**Comment by jpmnteiro** [comment]
Had the same problem after migrating a project from VS2015 to VS2017, TS project would not build on a build agent that only has the Build Tools. Ended up using the nuget package as suggested by @mhegazy (thank you).
However, it'd be nice to have a solution that didn't involve using a nuget package.

**Comment by MarcusFelling** [comment]
Thanks for the workaround @jpmnteiro. Just got done creating a build server with build tools only and ran into this issue.

**Comment by billti** [comment]
Build support for TypeScript has been added to the Build Tools SKU of Visual Studio 2017 since Update 6 (see https://docs.microsoft.com/en-us/visualstudio/releasenotes/vs2017-relnotes-v15.6#BuildTools ).

**Comment by davidgvh** [comment]
@billti, this appears to be an issue with Visual Studio 2022 Build Tools as well. Was there a regression?
